module/zfs/zil.c

```diff
@@ -1416,11 +1416,17 @@ zil_lwb_write_open(zilog_t *zilog, lwb_t *lwb)
  * aligned to 4KB) actually gets written. However, we can't always just
  * allocate SPA_OLD_MAXBLOCKSIZE as the slog space could be exhausted.
  */
-uint64_t zil_block_buckets[] = {
-    4096,		/* non TX_WRITE */
-    8192+4096,		/* data base */
-    32*1024 + 4096, 	/* NFS writes */
-    UINT64_MAX
+struct {
+	uint64_t	limit;
+	uint64_t	blksz;
+} zil_block_buckets[] = {
+	{ 4096,		4096 },			/* non TX_WRITE */
+	{ 8192 + 4096,	8192 + 4096 },		/* database */
+	{ 32768 + 4096,	32768 + 4096 },		/* NFS writes */
+	{ 65536 + 4096,	65536 + 4096 },		/* 64KB writes */
+	{ 131072,	131072 },		/* < 128KB writes */
+	{ 131072 +4096,	65536 + 4096 },		/* 128KB writes */
+	{ UINT64_MAX,	SPA_OLD_MAXBLOCKSIZE},	/* > 128KB writes */
```

Comment:
We did quite a bit of experimenting here a few years ago. In general, fewer buckets is better. But also the previous default buckets aren't representative of today's workloads.
Given the large size of disks today, especially for slogs, there doesn't seem to be any real space savings to be gained by having smaller buckets. Thus the only benefit for small buckets is if you have mixed workload datasets in raidz pools with no slog.

Reply:
Using NVDIMMs for SLOG we are quite space-constrained at this point.  May be new Apache Pass change that, but so far we are not there.